core/src/main/scala/kafka/coordinator/transaction/TransactionManager.scala

```diff
@@ -0,0 +1,300 @@
+    val partitionId = partitionFor(transactionalId)
+
+    loadingPartitions.contains(partitionId)
+  }
+
+  /**
+    * Gets the partition count of the transaction log topic from ZooKeeper.
+    * If the topic does not exist, the default partition count is returned.
+    */
+  private def getTransactionTopicPartitionCount: Int = {
+    zkUtils.getTopicPartitionCount(Topic.TransactionStateTopicName).getOrElse(config.numPartitions)
+  }
+
+  private def loadTransactionMetadata(topicPartition: TopicPartition) {
+    def highWaterMark = replicaManager.getHighWatermark(topicPartition).getOrElse(-1L)
+
+    val startMs = time.milliseconds()
+    replicaManager.getLog(topicPartition) match {
+      case None =>
+        warn(s"Attempted to load offsets and group metadata from $topicPartition, but found no log")
+
+      case Some(log) =>
+        val buffer = ByteBuffer.allocate(config.loadBufferSize)
+
+        val loadedPids = mutable.Map.empty[String, TransactionMetadata]
+        val removedPids = mutable.Set.empty[String]
+
+        // loop breaks if leader changes at any time during the load, since getHighWatermark is -1
+        var currOffset = log.logStartOffset.getOrElse(throw new IllegalStateException(s"Could not find log start offset for $topicPartition"))
+        while (currOffset < highWaterMark && !shuttingDown.get()) {
+          buffer.clear()
+          val fileRecords = log.read(currOffset, config.loadBufferSize, maxOffset = None, minOneMessage = true)
+            .records.asInstanceOf[FileRecords]
+          val bufferRead = fileRecords.readInto(buffer, 0)
+
+          MemoryRecords.readableRecords(bufferRead).entries.asScala.foreach { entry =>
+            for (record <- entry.asScala) {
+              require(record.hasKey, "Transaction state log's key should not be null")
+              TransactionLog.readMessageKey(record.key) match {
+
+                case txnKey: TxnKey =>
+                  // load pid metadata along with transaction state
+                  val transactionalId: String = txnKey.key
+                  if (record.hasNullValue) {
+                    loadedPids.remove(transactionalId)
+                    removedPids.add(transactionalId)
+                  } else {
+                    val txnMetadata = TransactionLog.readMessageValue(record.value)
+                    loadedPids.put(transactionalId, txnMetadata)
+                    removedPids.remove(transactionalId)
+                  }
+
+                case unknownKey =>
+                  throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+              }
+
+              currOffset = entry.nextOffset
+            }
+          }
+
+          loadedPids.foreach {
+            case (transactionalId, txnMetadata) =>
+              val currentTxnMetadata = addTransaction(transactionalId, txnMetadata)
+              if (!txnMetadata.equals(currentTxnMetadata)) {
```

Comment:
So I am probably being a bit thick here, but how is it possible to get into this case from the code in this PR? The `loadedPid` is a map from `transactionalId` to `txnMetadata`. When reading through the log, you will always retain _only_ the latest metadata for a given transactional id. And hence when iterating through the map entries, you should always encounter a single transactional id exactly once. 

In what cases will you encounter a single transactional id multiple times in the map, and hence mark a partition as corrupted?